File: src/command/codec/url.ts
import { invoke } from "@tauri-apps/api/core";

const encodeURL = (input: string) => {
  return invoke<string>("encode_url", { input });
};

const decodeURL = (input: string) => {
  return invoke<string>("decode_url", { input });
};

export { encodeURL, decodeURL };

File: src-tauri/src/command/codec/url.rs
use crate::command_error;

#[tauri::command]
#[tracing::instrument(level = tracing::Level::DEBUG, ret)]
pub fn encode_url(input: &str) -> String {
    urlencoding::encode(input).into_owned()
}

#[tauri::command]
#[tracing::instrument(level = tracing::Level::DEBUG, ret, err(level = tracing::Level::ERROR))]
pub fn decode_url(input: &str) -> Result<String, Error> {
    urlencoding::decode(input)
        .map(|v| v.into_owned())
        .map_err(|e| e.into())
}

command_error! {
    (Utf8, "utf8 error: {0}", #[from] std::string::FromUtf8Error),
}
